document std.algorithm.mutation: swapAt

greenify opened the pull request:
Hey I have seen that `swapAt` is used quite often even though it's undocumented.

In phobos `swapAt` is used 19 times in other modules and [other people](https://github.com/DlangScience/mir/pull/29) seem to know about it too.

Imho it looks nicer to write `myArray.swapAt(0, i)` than `swap(myArray[0], myArray[i])`.
I couldn't find a unittest for the  `else` in the `static if` - maybe @quickfur?

If you don't like to expose `swapAt` I would suggest we remove it entirely from Phobos.

```
algorithm/sorting.d:            swapAt(r, 0, len - 1);
algorithm/sorting.d:            swapAt(r, 0, mid);
algorithm/sorting.d:            swapAt(r, mid, len - 1);
algorithm/sorting.d:            swapAt(r, 0, mid);
algorithm/sorting.d:            swapAt(r, mid, len - 1);
algorithm/sorting.d:            swapAt(r, 0, mid);
algorithm/sorting.d:            swapAt(r, 0, len - 1);
algorithm/sorting.d:                swapAt(r, j, j + 1);
algorithm/sorting.d:        swapAt(r, pivotIdx, r.length - 1);
algorithm/sorting.d:            swapAt(r, lessI, greaterI);
algorithm/sorting.d:        swapAt(r, r.length - 1, lessI);
algorithm/sorting.d:            r.swapAt(0, i);
algorithm/sorting.d:                    r.swapAt(parent, child);
algorithm/sorting.d:            r.swapAt(parent, child);
algorithm/sorting.d:                    r.swapAt(parent, child);
algorithm/sorting.d:            r.swapAt(parent, child);
algorithm/sorting.d:            r.swapAt(parent, child);
container/binaryheap.d:            swapAt(_store, parentIdx, n);
random.d:        swapAt(r, i, uniform(i, r.length, gen));
```

9il commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2374,55 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
+    i1 = first index
+    i2 = second index
+*/
+void swapAt(R, I)(R r, I i1, I i2)
```
Why not just `size_t`

greenify replied:
It actually was `size_t` in the beginning - the generic behavior was "copied" from `moveAt`.
Sometimes a library by many people can be confusing - I changed it back to `size_t`

9il replied:
We need the same style like `moveAt` :-/... This was just a question) Please add `in` for generic integral type: this will reduce template bloat

greenify replied:
> This was just a question

Never mind ;-)

> We need the same style like moveAt

Well the current `swapAt` in Phobos uses `size_t`, so probably no one would complain if we keep it like this?

Hackerpilot commented on std/algorithm/mutation.d:
```diff
@@ -51,6 +51,8 @@ $(T2 stripRight,
         $(D stripRight!(e => e == 1)(a)) returns $(D [1, 1, 0]).)
 $(T2 swap,
         Swaps two values.)
+$(T2 swap,
```
Should this be `swapAt` to match the function name?

greenify replied:
Damn copy & paste. Thanks for spotting.

9il commented:
Small API review is required.

tsbockman commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2374,52 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
+    assert(a.equal([1, 2, 3]));
+
+    a.swapAt(1, 1);
+    assert(a.equal([1, 2, 3]));
```
Why is this test duplicated?

DmitryOlshansky commented:
LGTM the usefulness is pretty low. Personally I think it was a local helper that was accidentally made public.

9il commented:
> Personally I think it was a local helper that was accidentally made public.

@andralex uses this function in `std.algorithms` with UFCS syntax. The goal of this is that user defined range can override `swapAt` with method to achieve better perfomance if it is possible.

9il commented:
@greenify , please change `swapAt` usage in Phobos with UFCS (in this PR)

greenify commented:
> @greenify , please change swapAt usage in Phobos with UFCS (in this PR)

done.

tsbockman commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2374,49 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
+    i1 = first index
+    i2 = second index
+*/
+void swapAt(R)(R r, size_t i1, size_t i2)
```
How about a template constraint, to match the docs?
```D
if (isRandomAccessRange!R)
```

9il replied:
No, please, `swapAt` would work even for non random access ranges.

tsbockman replied:
@9il 
> No, please, `swapAt` would work even for non random access ranges.

Then why are you merging this? Shouldn't the docs be fixed?

tsbockman replied:
What are the actual requirements for `R`?

9il replied:
yes, @greenify please fix the docs

DmitryOlshansky replied:
I guess it still makes a random-access range, I'd be surprised to see range that swaps with arbitrary indices but provides no RA.

The point is rather to support sealed ranges, that return rvalues from opIndex.

9il replied:
@DmitryOlshansky see `moveAt` for example: https://github.com/D-Programming-Language/phobos/blob/75702dc9957996d4cdb87c65b3104ac8d0ea5e67/std/range/primitives.d#L1946

It has not any constaints

tsbockman replied:
Studying this more closely... some types that should work as `R`, but which do not meet the formal definition of a "*random access range*":

- `char[]` and `wchar[]`, because of auto-decoding
- Static arrays
- Custom types that implement `opIndex()` to return an *lvalue*
- Custom types that have a `moveAt()` method *and* implement `opIndex()` with a return type that overloads `opAssign()` or has an assignable `alias _ this` - even if `R`'s `opIndex()` returns an *rvalue*

So, we can either write an elaborate explanation of the true requirements, or just change `a random access range` to `a range` and let people guess-and-check / read the source code.

9il replied:
just `a range`

9il commented:
Auto-merge toggled on

9il commented:
Auto-merge toggled off

9il commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2374,49 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
+    else
+    {
+        if (i1 == i2) return;
+        auto t1 = moveAt(r, i1);
```
This should be UFCS too

9il replied:
Do not change it in this PR. It is already changed in #4165

greenify commented:
> yes, @greenify please fix the docs

Done.

> Studying this more closely... some types that should work as R, but which do not meet the formal definition of a "random access range":

Added them all as unittests - it turns out that for static arrays `ref` is needed. 
I also added a small test to `swap` for static arrays.

9il commented:
`auto ref please`

9il commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2378,120 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
+        }
+    }
+    auto f = RandomMoveAtType!(int[])([1, 2, 3]);
+    f.swapAt(1, 2);
```
This should be not UFCS

greenify commented:
> auto ref please

thanks and rebased, so that the first commits contains all `swapAt` changes and the second one changes UFCS usage of swapAt within Phobos.

tsbockman commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2378,120 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
 }
 
+/**
+Swaps two elements in-place of a random access range `r`,
```
@greenify By "fix the docs", we meant to change `a random access range` here to just `a range`, because `random access` wrongly implies that `isRandomAccessRange!R` is required.

greenify replied:
Sorry forgot that one - I was too focused on adding all the unit tests.

tsbockman commented:
When placing several very similar tests within the same unit test block, it is a good idea *not* to reuse the same exact test values (like `[1, 2, 3]`) over and over. The problem is that it's far too easy to do this:
```D
unittest
{
    auto v1 = foo!byte(1, 1);
    assert(v1 == 2);
    auto v2 = foo!ubyte(1, 1);
    assert(v2 == 2);
    auto v3 = foo!short(1, 1);
    assert(v3 == 2);
    auto v4 = foo!ushort(1, 1);
    assert(v3 == 2);
    auto v5 = foo!int(1, 1);
    assert(v5== 2);
    auto v6 = foo!uint(1, 1);
    assert(v6 == 2);
}
```
In the above, `v4` is never tested.

(I just caught a bug like this in another pull request I was reviewing earlier this week.)

greenify commented:
> When placing several very similar tests within the same unit test block, it is a good idea not to reuse the same exact test values (like [1, 2, 3]) over and over. 

Understood - thanks for the hint. Rebased.

tsbockman commented:
@9il This is ready now, I think?

9il commented:
Auto-merge toggled on

9il commented:
Thank you @greenify

tsbockman commented on std/algorithm/mutation.d:
```diff
@@ -2389,6 +2378,120 @@ void swap(T)(ref T lhs, ref T rhs) if (is(typeof(lhs.proxySwap(rhs))))
+
+    int[3] c = [1, 2, 3];
+    c.swapAt(1, 2);
+    assert(c.array.equal([1, 3, 2]));
```
@greenify You're missing an import for `std.array.array()` here; that's why it's failing the auto-tester.

(Probably caused by https://github.com/D-Programming-Language/phobos/pull/4169 ?)

greenify commented:
> @greenify You're missing an import for std.array.array() here; that's why it's failing the auto-tester.

Added and Rebased. Thanks for the ping.

> (Probably caused by #4169 ?)

Most likely.
